### PR TITLE
Final shape / free func tweaks

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
- [x] `sample`
- [x] `isoline`
- [x] detailed `check` output
- [x] expose API in `func`: `from cadquery.func import *`
- [x] tests

afbeelding
afbeelding

### CODE REVIEW

File: tests/test_shapes.py [hunk 1]
```diff
@@ -136,3 +138,36 @@ def test_bin_import_export():
...
+
+    e = ellipse(10, 1)
+    s = segment((0, 0), (1, 0))
+
+    pts1, params1 = e.sample(10)  # equidistant
+    pts2, params2 = e.sample(0.1)  # deflection based
+    pts3, params3 = s.sample(10)  # equidistant, open
+
+    assert len(pts1) == len(params1)
+    assert len(pts1) == 10 - 1  # e is closed
```
Comment (lorenzncode) [inline tests/test_shapes.py:153]: I see this is not an off by one error but by design (the difference in count for closed/open edges).

I compared with `Sketch`.  `Sketch` has different handling here for closed edges and wires.

```py
import cadquery as cq
from cadquery.func import *
from cadquery.vis import show

# sample 4
wire1 = circle(20)
(pos, params) = wire1.sample(4)
samples = compound([sphere(4).moved(p) for p in pos])

# compare with sketch distribute
sk1 = cq.Sketch().circle(30).edges().distribute(4).circle(2)

show(wire1, samples, sk1)
```

I just want to point out this difference to confirm the design.

Comment (adam-urbanczyk) [reply tests/test_shapes.py]: Good point, I fixed it to be consistent with `Sketch`.

File: cadquery/occ_impl/shapes.py [hunk 2]
```diff
@@ -1914,7 +1925,8 @@ def positionAt(
     def positions(
         self: Mixin1DProtocol, ds: Iterable[float], mode: ParamMode = "length",
     ) -> List[Vector]:
-        """Generate positions along the underlying curve
+        """
+        Generate positions along the underlying curve
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:1929]: Perhaps add punctuation to the end of this description sentence.  Compare with positionAt.  I know it's  not consistent everywhere.

File: cadquery/occ_impl/shapes.py [hunk 3]
```diff
@@ -1973,7 +2015,8 @@ def locations(
         frame: FrameMode = "frenet",
         planar: bool = False,
     ) -> List[Location]:
-        """Generate location along the curve
+        """
+        Generate location along the curve
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:2019]: punctuation

**Review by lorenzncode — APPROVED** [review]
I like the new free func import method!

File: cadquery/occ_impl/shapes.py [hunk 4]
```diff
@@ -1923,14 +1935,44 @@ def positions(
 
         return [self.positionAt(d, mode) for d in ds]
 
+    def sample(
+        self: Mixin1DProtocol, n: Union[int, float]
+    ) -> Tuple[List[Vector], List[float]]:
+        """
+        Sample a curve based on a number of points or deflection.
+
+        :param n: number of positions or deflection
```
Comment (jmwright) [inline cadquery/occ_impl/shapes.py:1944]:
```suggestion
        :param n: Number of positions or deflection
```

File: cadquery/occ_impl/shapes.py [hunk 5]
```diff
@@ -1923,14 +1935,44 @@ def positions(
...
+        :return: A list of Vectors and a list of parameters.
```
Comment (jmwright) [inline cadquery/occ_impl/shapes.py:1945]: Isn't this supposed to be "returns" in this context?

**Review by jmwright — APPROVED** [review]
Thanks @adam-urbanczyk !

File: cadquery/occ_impl/shapes.py (hunk 5 above)
Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py:1945]: According to the docs `:return:` https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html

**Pull request merged**